src/planner/explain.rs

```diff
@@ -65,210 +83,261 @@ impl<'a> Explain<'a> {
     #[inline]
     fn is_true(&self, id: &Id) -> bool {
         self.expr[*id] == Expr::true_()
     }
-}
 
-impl Display for Explain<'_> {
-    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
+    pub fn pretty(&self) -> Pretty<'a> {
         use Expr::*;
         let enode = &self.expr[self.id];
-        let tab = self.tab();
         let cost = self.cost();
         match enode {
-            Constant(v) => write!(f, "{v}"),
-            Type(t) => write!(f, "{t}"),
+            Constant(v) => Pretty::display(v),
+            Type(t) => Pretty::display(t),
             Table(i) => {
                 if let Some(catalog) = self.catalog {
-                    write!(f, "{}", catalog.get_table(i).expect("no table").name())
+                    catalog.get_table(i).expect("no table").name().into()
                 } else {
-                    write!(f, "{i}")
+                    Pretty::display(i)
                 }
             }
             Column(i) => {
                 if let Some(catalog) = self.catalog {
-                    write!(f, "{}", catalog.get_column(i).expect("no column").name())
+                    let column_catalog = catalog.get_column(i).expect("no column");
+                    column_catalog.into_name().into()
                 } else {
-                    write!(f, "{i}")
+                    Pretty::display(i)
                 }
             }
-            ColumnIndex(i) => write!(f, "{i}"),
-            ExtSource(src) => write!(f, "path={:?}, format={}", src.path, src.format),
-            Symbol(s) => write!(f, "{s}"),
+            ColumnIndex(i) => Pretty::display(i),
 
-            Ref(e) => write!(f, "{}", self.expr(e)),
-            List(list) => {
-                write!(f, "[")?;
-                for (i, v) in list.iter().enumerate() {
-                    if i != 0 {
-                        write!(f, ", ")?;
-                    }
-                    write!(f, "{}", self.expr(v))?;
-                }
-                write!(f, "]")
-            }
+            // TODO: use object
+            ExtSource(src) => format!("path={:?}, format={}", src.path, src.format).into(),
+            Symbol(s) => Pretty::display(s),
+            Ref(e) => self.expr(e).pretty(),
+            List(list) => Pretty::Array(list.iter().map(|e| self.expr(e).pretty()).collect()),
 
             // binary operations
             Add([a, b]) | Sub([a, b]) | Mul([a, b]) | Div([a, b]) | Mod([a, b])
             | StringConcat([a, b]) | Gt([a, b]) | Lt([a, b]) | GtEq([a, b]) | LtEq([a, b])
             | Eq([a, b]) | NotEq([a, b]) | And([a, b]) | Or([a, b]) | Xor([a, b])
-            | Like([a, b]) => write!(f, "({} {} {})", self.expr(a), enode, self.expr(b)),
+            | Like([a, b]) => named_record(
+                enode.to_string(),
+                btreemap! {
+                    "lhs" => self.expr(a).pretty(),
+                    "rhs" => self.expr(b).pretty(),
+                },
+                vec![],
+            ),
```

Comment:
I agree, but it's quite complicated... It would be nice if we can do a precedence-driven pretty printing, but that's another story and requires a major rewrite of the data structure (I believe)

Reply:
My crate simply does not handle these simple cases, because inline expressions should be separated from larger data structures, for instance you don't want `Add(Record(), Scan())` to be a valid expression, and it would be nice if we can rule this monstrosity out by the type system..